qcodes/interactive_widget.py

```diff
@@ -0,0 +1,436 @@
+    _update_nested_dict_browser(nested_keys, nested_dict, box)(None)
+    return box
+
+
+def _plot_ds(ds: DataSet) -> None:
+    try:
+        # `get_data_by_id` might fail
+        nplots = len(get_data_by_id(ds.run_id))  # TODO: might be a better way
+        nrows = math.ceil(nplots / 2) if nplots != 1 else 1
+        ncols = 2 if nplots != 1 else 1
+        fig, axes = plt.subplots(nrows, ncols, figsize=(6 * ncols, 4 * nrows))
+        # `plot_dataset` might also fail.
+        plot_dataset(ds, axes=axes.flatten())
+        fig.tight_layout()
+        plt.show(fig)
+    except Exception as e:
+        print(e)  # TODO: print complete traceback
+
+
+def _do_in_tab(tab: Tab, ds: DataSet, which: str) -> Callable[[Button], None]:
+    """Performs an operation inside of a subtab of a `ipywidgets.Tab`.
+
+    Args
+        tab: Instance of `ipywidgets.Tab`.
+        ds: A qcodes.DataSet instance.
+        which: Either "plot" or "snapshot".
+    """
+
+    def delete_tab(output: Output, tab: Tab) -> Callable[[Button], None]:
+        def on_click(_: Button) -> None:
+            tab.children = tuple(c for c in tab.children if c != output)
+
+        return on_click
+
+    def _on_click(_: Button) -> None:
+        assert which in ("plot", "snapshot")
+        title = f"RID #{ds.run_id} {which}"
+        i = next(
+            (i for i in range(len(tab.children)) if tab.get_title(i) == title),
+            None,
+        )
+        if i is not None:
+            # Plot/snapshot is already in the tab
+            tab.selected_index = i
+            return
+        out = Output()
+        tab.children += (out,)
+        i = len(tab.children) - 1
+        tab.set_title(i, title)
+        with out:
+            clear_output(wait=True)
+
+            close_button = button(
+                f"Close {which}",
+                "danger",
+                on_click=delete_tab(out, tab),
+                button_kwargs=dict(icon="eraser"),
+            )
+            display(close_button)
+
+            try:
+                if which == "plot":
+                    _plot_ds(ds)
+                elif which == "snapshot":
+                    snapshot = ds.snapshot
+                    if snapshot is not None:
+                        display(nested_dict_browser(snapshot))
+                    else:
+                        print("This dataset has no snapshot")
+            except Exception as e:
+                print(e)  # TODO: print complete traceback
+        tab.selected_index = i
+
+    return _on_click
+
+
+def create_tab(do_display: bool = True) -> Tab:
+    """Creates a `ipywidgets.Tab` which can display outputs in its tabs."""
+    tab = Tab(children=(Output(),))
+
+    tab.set_title(0, "Info")
+    if do_display:
+        display(tab)
+
+    with tab.children[-1]:
+        # Prints it in the Output inside the tab.
+        print("Plots and snapshots will show up here!")
+    return tab
+
+
+def editable_metadata(ds: DataSet) -> Box:
```

Comment:
note that due to the value the `DataSet.metadata` is implemented, changes to the notes made in the widget would be persistet but won't be visible to any other isntances of the same dataset elsewhere unless that dataset gets reloaded. This is just a note, it's unrelated to the widget but related to the `DataSet.metadata` feature which qcodes will improve at some point.

Reply:
OK, makes sense. I don't think I could do anything about it really. There is an explanation in the `experiments_widget`'s doc-string.